### ISSUE TITLE
Roadmap for inclusion of GPU implementation of matrix free in Deal.II

### ISSUE THREAD

**Issue by kalj** [dealii/dealii#2351]
During the last 18 months or so, I've been implementing something equivalent to the matrix free functionality but for execution on a GPU. With good preliminary performance results, and a somewhat decent functionality set, I think it is now time to start sketching on how this should be merged into Deal.II.

After some discussions with @bangerth, I have come to the following conclusions:

a) This should be merged into the existing data structures as much as possible, to achieve nice encapsulation and avoid code duplication.
b) It should be split up in as small independent pieces as possible, to facilitate the patch review process.

Luckily, since the code has been designed to mimic the CPU code as closely as possible, this should not be all that difficult.

The code has been implemented as an application in "user space" outside the Deal.II library itself, but with most of the new infrastructure in a subdirectory. It currently consists of three major parts:

VectorGpu

This corresponds to a dealii::Vector<>, and is an object used from the CPU but with internal data on the GPU and code executed on the GPU. This could quite naturally be merged with the Vector class if we include a template parameter which decides the internal storage type. I think that, after all, they have to be different types, since it does not make sense to use a GPU Vector in all cases that we can use a Vector. For instance, access of individual entries from the CPU using operator() is not allowed to a GPU Vector (of course it can be implemented but it shouldn't really be used for performance reasons). Another difference is that the GPU version needs a getData() function which returns a pointer to the internal (GPU-side) data. I think we need this for the user to be able to initialize arrays, like variable coefficients.

MatrixFreeGpu

This corresponds to a dealii::MatrixFree and sets up the internal data structure caches used in the evaluation. It also contains a function cell_loop which is similar to the regular cell_loop, but takes its local operation packed in a struct since function pointers to GPU-side code cannot be passed with CUDA. Also, having a struct allows for passing of additional problem-dependent data, like a variable coefficient. I think we should be able to include this too as a variant of MatrixFree, again using a template parameter to distinguish between them. The main complication here is the interface to the cell_loop function, where we need to agree on something that is general enough to work for both CPU and GPU. I think we should be able to. What I need for the GPU is to be able to pass something with a function and some (read-only) data, and passing a member function or a function pointer is not allowed in CUDA (for the time being). We could have a convention that the struct to be passed must implement a "local_apply" function. We could even enforce that by creating an abstract base "LocalOperation" which has to be derived. Also, on the GPU, this function cannot take the MatrixFree object itself as a parameter since that is a CPU-side object, so that's something we have to allow for. Note that we could still offer the existing function-pointer interface as an overload for the CPU version.

FEEvaluationGpu

This is the most distinct part from it's corresponding CPU version. This now has to be an entirely GPU-side object, which means that it has to be free of all CPU-only code. It also has a slightly different interface since it is created once for each thread, and has quite distinct implementations of the operations. Because of this, there is little overlap between the two. Here, I think it will be a significant challenge to create a merged version. This is a very performance critical class, so it is vital that we keep it as lightweight as possible.

Build setup

Apart from these classes, one needs to decide on an infrastructure for handling compilation. Cmake supports CUDA, so it is quite straightforward to set everything up. Currently, I include Deal.II manually, but we should be able to include something for CUDA in the autopilot framework. Probably, the most natural way would be to have something like DEAL_II_WITH_CUDA which enables the GPU functionality. Without this, nothing is needed in terms of GPU, CUDA framework, etc. If enabled, CUDA and a suitable compatible backend C++ compiler should be set up (usually  GCC >=4.9 is good, but the Intel compiler "should" work too). Since we change compiler to the CUDA nvcc one, I'm sure this will give lot's of interesting warnings for all the regular CPU code, but we can try. :) In theory it should behave as the backend compiler for all CPU code.

Some questions: 
1. Any thoughts on the idea of using a template parameter to distinguish between the two implementations? Of course, we can have a default value there, which equals to the existing CPU version, and not affect user code significantly.
2. What do you think of my idea for a new interface for cell_loop?
3. Any suggestions on how to deal with FEEvaluation? Can we have two quite different classes combined into one to the user without too much overhead?
4. How about the build setup idea? I like it because it should make everything non-intrusive in terms of not being visible if not enabled, and therefore not break things.

Being the master of the matrix free implementation, @kronbichler, what are your thoughts on this?

I should be able to show you a cleaned up and commented version of the code within a couple of days.

**Comment by Rombur** [comment]
VectorGpu might be more complicated than you think :-/ We are rewriting the vector interface and VectorGpu will have to derive from VectorSpaceVector or ReadWriteVector (the new dealii::Vector derives from both https://dealii.org/developer/doxygen/deal.II/classLinearAlgebra_1_1Vector.html) and I was told that CUDA doesn't play well with virtual function... The other thing is that while you should use the new interface, it isn't totally ready yet. So you will need some work-around until we finish it.

**Comment by kalj** [comment]
That's nice -- it's always good with interfaces. It should definitely derive from VectorSpaceVector, and maybe ReadWriteVector too, but again, direct access of entries from the "outside" shouldn't really be done. Virtual function is not a problem in this case since this will be a CPU class with CUDA internals. It may be an issue for classes that live on the GPU themselves. So it would affect the way that we specify an interface for the local operation, so we will probably have to go with static polymorphism (templates) but that's is not really a restriction.

**Comment by bangerth** [comment]
I think the GPU vector should be class of its own, rather than selected via a template argument. It serves a separate purpose, so deserves a name of its own. For reference, the github issue that discusses everything related to vectors is #745.

**Comment by kalj** [comment]
@bangerth: That makes a lot of sense, it should be a distinct type after all.

Here is now a relatively cleaned up version of my code as of now: https://github.com/kalj/dealii-cuda
I added some comments, but it is still far from having thorough documentation. (@Rombur: this is the same repo, but will be used from now on).

**Comment by bangerth** [comment]
@kalj -- it's difficult to elucidate the difference between entire repositories. But I know you're already working on breaking the differences into individual patches :-)

**Comment by guidokanschat** [comment]
@kalj: Out of curiosity: from my experience with GPU, you should avoid uploading the whole deal.II library because of its size. Did you start separating local from global code?

**Comment by kalj** [comment]
@guidokanschat, thanks for your concern, but I feel that I don't fully understand the question, perhaps you can reformulate it?

The idea is to implement GPU versions of the main building blocks used in a matrix-free operator application, i.e. mainly the FEEvaluation class. Then, only the operations that the user actually uses in their operator will need to be compiled into the executable. As for data, we precompute everything (transformation Jacobians, variable coefficients, shape_values, etc) on the CPU, and then copy it over to GPU memory. Finally, all "variable data", e.g. solution vectors, are kept on the GPU throughout the whole solver, so we avoid frequent data transfer over the lousy PCIe bus. To my understanding, this should be quite efficient, and I don't see how the fact that we have a large library surrounding this could be a problem.

**Comment by bangerth** [comment]
I think what @guidokanschat was saying is this: compiling the *entire* library to run on GPUs may not be very useful since it's such a huge piece of code that would take up a sizable amount of memory if it has to replicated multiple times for each thread/task/whatever the right terminology is on the GPU. 

I don't know whether that's even possible, but it can for example be done for Xeon Phi cards.

**Comment by kronbichler** [comment]
I like the general design of the classes and the layout. Regarding some of the details, I have a few questions of course:

1. Regarding templated objects for vectors and MatrixFreeGpu: I think that we should strive to make it relatively easy to switch between a regular CPU-based evaluation and a GPU version. The point is that I would like to write a common code that implements various differential operators, test on the CPU and then run the same code on the GPU. If you have a somewhat complex code, you can easily have 10-30 `local_apply` functions, including operations that do face integrals. You don't want to re-write and re-test all those functions when going to the GPU, so you would ideally just put one or a few typedefs to a different value. I think that this should be easy enough if we have different class names such as VectorGpu and MatrixFreeGpu as long as the functional interface is the same. The Trilinos people have decided to go with templates (if you check the 'Tpetra' code base, almost all classes have a 'Node' template argument that transparently selects whether to use CUDA, OpenMP and additional tricks such as memory allocation on the fast device memory of Knights Landing Xeon Phis rather than the global RAM pool), but it is certainly overkill for us to do the same at this point. Thus, I would agree with @bangerth that we can keep things separated right now. (Things might change if we need different back-ends for CPU vs. GPU but there is a large enough common layer on top of that which should have the same code for both.)

2. Changing the interface to cell_loop (or a more generic loop once you want to invoke not just cell integrals but also face and boundary integrals) is certainly possible. I checked the MatrixFreeGpu::cell_loop in your code (in matrix_free_gpu.h) and have a few remarks:
  - If I want to have different integrators in my code, e.g. one that computes a matrix-vector product and another one that computes a residual, I guess I have to put them in separate structs? What restrictions does CUDA pose on those separate structs? A typical situation in nonlinear problems is that the function that computes the residual also computes a 'linearization point' that is written into an array (for each cell and quadrature point) that later gets read in the matrix-vector product. Naturally, the two functions need to share the array with the linearized data. Another application I often have are stabilization parameters that you would like to re-use in different local_apply functions. While I could copy them to each evaluator, it seems quite expensive.
  - Do we need to expose the 'color' to the user space or could this be encapsulated inside the cell loop before you invoke apply_kernel_shmem?
  - In MatrixFree, there is a template on the vector type because you might have evaluators that take several vectors. One example is time stepping such as BDF-2 where you need `u^{n-1}` and `u^{n-2}` in the expression for the old solution. To solve this, we put several vectors into a std::vector<VectorType> (or std::vector<VectorType*>), possibly doing some stuff on those vectors (import ghost values in MPI mode), and passing them along to the user's local_apply function. Is something similar possible with your code?

3. I guess that keeping FEEvaluation separate is fine as long as you can select them via a typedef or something similar to allow for easily porting the code from one architecture to the other. Certainly, the implementation details are sufficiently different at this point (and efficiency mandates that we do it the right way in both cases) that we need different back-ends but for the user the change should be merely to select some 'use_cpu' or 'use_cuda' without having to think about it. On the other hand, I wonder if we can re-use some of the internal loops/functions in both code: The implementation in `matrix_free_gpu/tensor_ops.cuh` looks not too different from what we use for the CPU version if we abstract away the `__device__` keyword and `__syncthreads()` and forget about all the special cases in the CPU version. We do not need to solve this now but I like to keep the code close in order to make both implementations profit from each other. For example, the CPU version has some specializations for various types of elements ('symmetric' basis functions including even-odd decomposition (I can explain that if you want), truncated tensor products, tensor products + enrichment) that enter different functions. You would like to enable those on a GPU as well at some point I guess.

As you suggest, we should try to get to similar interfaces with the CPU-based version and we can definitely make changes to the CPU version (we have work in the pipeline for face integrals in DG that should be ready in a few months). In particular, I see that your model mainly expects an operation on the quadrature points. Maybe we could make a joint effort to provide interfaces that free the user from writing the read_dof_values(), evaluate(), integrate() and distribute_local_to_global() functions in the usual case. This should allow for merging some `__syncthreads()` I could imagine and in particular hide those calls from the user code. I'm not sure how far we can go because one can imagine operations that need to have access to something in between the quadrature points - think of a stabilization parameter that is computed as the maximum of some solution quantity over all local quadrature points - you would need to pass through quadrature points once for finding the maximum, and then once again for compute the local residual.

**Comment by kalj** [comment]
Thanks for great comments, @kronbichler!

1. Then I think we all agree that we can have separate classes that behave the same. In some cases, such as for the vector, it is quite natural with a common interface (abstract base class), but for MatrixFree, and FEEvaluation in particular, we should probably go with "duck typing" / static interface.

2. 
  - (a) I see the problem. In CUDA, you can pass non-member functions as pointers, and you can pass structs like I do, but you cannot pass member functions. I guess with a class-imitating system where the function takes the object as it's first parameter, we could get the same behavior, but it is not very nice. We could also have several functions in the same struct to share data, but then we couldn't use a reserved function name "apply". If we would create one struct per local_apply, we could still share data with pointers to common data. Then no data is copied so it shouldn't have any performance penalty, but I agree that it is not as nice as a single struct with different functions.
  - (b) This sure is desirable. Right now, the user doesn't know about the number of colors, only that they are there. I think it may be difficult to hide, since stuff like the computation of coefficients should be implemented by the user, but the storage pattern depends on the coloring. Now that I think of it, I must say that, for the parallelization with 1thread/DoF, I can't see why it wouldn't work since we don't care about the order of elements. :) But if we could hide both the storage pattern and the parallelization details (e.g. coloring) that would be the best.
  - (c) Do I understand you correctly if you need cell_loop and the local_apply to operate on a `std::vector<Vector>` (of compile-time-known length) rather than a plain `Vector`s? That's not implemented ATM, but I can't see how it could be a problem if we could just make that information available at compile time to MatrixFreeGpu, to set up the right functions. The local_apply function right now takes data pointers, but I guess we could wrap those up in some simple struct that could be templated. The problem is that this function takes pure GPU-side data, so GpuVector and similar is out of the question, but nothing stops us from creating and using said GPU-side struct. I will think some more about this.

3. I agree that the 'first level' of tensor operations, the ones which just calls the one dimensional reductions, are quite similar so there we can probably share some code. However, the one-dimensional reductions aren't really all that similar in my opinion as they are done in parallel for the GPU. But I guess this is all internal details, so it is not so crucial at this point. I'd like to keep it simple for now, but I definitely agree that those kinds of extensions should be added further down the road.

4. As a start, I think we should be able to make apply/local_apply look fairly identical to the CPU one. The reason for explicit `__syncthreads()` inside user code is that sometimes it is not necessary (e.g. when operating on only thread-local data), but I think we sacrifice very little generality/performance by moving them into FEEvaluationGPU. As for the `apply_quad_point_operations`, I think it is a fairly general and nice way of hiding how/if we loop over local points. If we could let the user call it several times with different callbacks, I think it could cover your stabilization example. But yeah, most of the rest of the local_apply function is very standard and maybe a more streamlined interface seems desirable. I recall the discussion we had about expression templates which I thought seemed like a nice approach.
I think one can imagine a two-level interface where there are high-level constructs for the common case (e.g. only provide the quadrature-point-local operation), and then low-level constructs (essentially the existing ones), for more involved stuff.

**Comment by kronbichler** [comment]
> 2.a [...] If we would create one struct per local_apply, we could still share data with pointers to common data.

I think that creating a struct with unique local_apply function would be an acceptable solution. It requires to set up a struct for each loop that one wants to run but it is not overly restrictive otherwise. The CPU version might still support function pointers but we could advertise the solution with the structs. Regarding names, I would suggest to use `apply_cell` as the name of the function (to allow for a future extension with `apply_face` and `apply_boundary` functions on faces).

> 2.b [...] But if we could hide both the storage pattern and the parallelization details (e.g. coloring) that would be the best.

Do you want to think about it a bit more? I think that it would indeed be nice to avoid those vectors. Regarding the pointers: If the user fills all coefficient data and other stuff through the same interface, we should be able to have the right layout internally without exposing the additional vector to the user, shouldn't we?

> 2.c Do I understand you correctly if you need cell_loop and the local_apply to operate on a std::vector<Vector> (of compile-time-known length) rather than a plain Vector?

Since it's a std::vector, it is actually not really necessary to know the stuff at compile time for the loop part. However, the evaluators inside need to know the length of the vector so essentially this is a compile-time constant.

> 4. I think we should be able to make apply/local_apply look fairly identical to the CPU one. [...] I think one can imagine a two-level interface where there are high-level constructs for the common case (e.g. only provide the quadrature-point-local operation), and then low-level constructs (essentially the existing ones), for more involved stuff.

This sounds like a great plan. Let me know when you have moved forward. I would agree that putting `__syncthreads()` into the FEEvaluationGpu calls is a good solution. Regarding expression templates or similar concepts I think we can keep that in mind but it is something we can do at a later stage.

**Comment by kalj** [comment]
> I think that creating a struct with unique local_apply function would be an acceptable solution. [...] Regarding names, I would suggest to use `apply_cell` as the name of the function

Great. Then, lets decide on this.

> Do you want to think about it a bit more? I think that it would indeed be nice to avoid those vectors. Regarding the pointers: If the user fills all coefficient data and other stuff through the same interface, we should be able to have the right layout internally without exposing the additional vector to the user, shouldn't we?

Do you mean that we could offer functionality for assembly of coefficients and other vectors and thus choose the layout? That seems like the way to go.

In my opinion, ideally, we would want the user to be able to be as free as possible in terms of what data could be used, but still decide internally how to layout that data. Maybe we can figure out a number of typical access patterns and create corresponding assembly functions? E.g. "access data per quadrature point", maybe you know of others? For a advection problem, I guess one gets a vector valued coefficient, and to be able to reshuffle dimensions (components, quad points, cells) is necessary.

In any case, to choose the right layout for arrays without any knowledge of how they will be accessed seems like an impossible task.

> Since it's a std::vector, it is actually not really necessary to know the stuff at compile time for the loop part. However, the evaluators inside need to know the length of the vector so essentially this is a compile-time constant.

I still feel a bit confused about the situation. If I understood correctly, we want to allow for
```
template <typename LocalOpT>
MatrixFree::cell_loop(std::vector<GpuVector> &out_vec,
                      const std::vector<GpuVector> &in_vec,
                      const LocalOpT &lop) const;
template <...>
class LocalOp {
  void apply_cell(gpu_array<n> out, const gpu_array<n> in, ... )
  {
    FEEvaluation phi(..);
    phi.read_dof_values(in.vec[0]);
    ...
  }
}
// where
template <int n>
struct gpu_array {
  Number *vec[n];
};
```
Now, what we need here is a way to convert between `out_vec` and `out`, which means we need to let `cell_loop` know the length at compile time. I would even argue that `std::tuple` (or some other array type with constant length) would be a more appropriate type for `out_vec` here. After all, we don't expect them to change in length, right?

**Comment by drwells** [comment]
Can we close this? We have made a lot of progress towards GPU support in the last four years.
